lib/asset/baseAsset.js

```diff
@@ -0,0 +1,296 @@
+    }
+
+    /**
+     * Pass in an array of ID strings for "sets" (categories, tags). The ID
+     * string is not the human-readable string (though it's possible for the
+     * two to be identical); you can create a human-readable name for the set
+     * in your app using other tools.
+     *
+     * Compare {@link NewsArticle#set_section}.
+     *
+     * @param {string[]} value - List of ID strings
+     */
+    set_tags(value) {
+        this._tags = value;
+    }
+
+    _normalize_tags() {
+        return this._make_tags().concat(this._ekn_tags)
+            .filter(tag => typeof tag === 'string')
+            .map(tag => tag.trim());
+    }
+
+    /**
+     * @returns {(Buffer|null)}
+     */
+    to_data() {
+        return null;
+    }
+
+    /**
+     * @returns {Object}
+     */
+    to_metadata() {
+        // Build an object with all required fields
+        const metadata = {
+            'assetID': this._asset_id,
+            'objectType': this._object_type,
+            'contentType': this._content_type,
+
+            'canonicalURI': this._canonical_uri,
+            'matchingLinks': [this._canonical_uri],
+
+            'title': this._title,
+            'tags': this._normalize_tags(),
+            'revisionTag': getTimestamp(),
+        };
+
+        // Add optional fields if they're present
+        if (this._date_published) {
+            metadata.published = this._date_published.toISOString();
+        }
+        if (this._last_modified_date) {
+            metadata.lastModifiedDate = this._last_modified_date.toISOString();
+            metadata.revisionTag = this._last_modified_date.toISOString();
+        }
+        if (this._sequence_number !== undefined) {
+            metadata.sequenceNumber = this._sequence_number;
+        }
+        if (this._license) {
+            metadata.license = this._license;
+        }
+        if (this._synopsis) {
+            metadata.synopsis = this._synopsis;
+        }
+        if (this._thumbnail_asset_id) {
+            metadata.thumbnail = this._thumbnail_asset_id;
+        }
+
+        return metadata;
+    }
+
+    _process() {
+        return Promise.resolve();
+    }
+
+    _make_tags() {
+        return this._tags;
+    }
+
+    _replace_dom_asset_ids() {
+        const $elem = this._document;
+
+        if (!$elem) {
+            throw new Error('Must call asset.render() before saving ' +
+                            `${this.constructor.name} to a hatch.`);
+        }
+
+        $elem('[data-libingester-asset-id]').each(function () {
+            const $asset = cheerio(this);
+
+            const assetId = $asset.attr('data-libingester-asset-id');
+            if (assetId) {
+                $asset.attr('data-soma-job-id', assetId);
+                $asset.attr('data-libingester-asset-id', null);
+            }
+
+            if (!$asset.attr('data-soma-job-id')) {
+                throw new ValidationError('Document has media without associated asset. ' +
```

Comment:
I don't see why this validation is not part of the validation.js module. But we can change that later.

Reply:
Nothing, I confused `ValidationError` with `VerificationError`. They look similar.